crypto/x509/by_file.c

```diff
@@ -89,22 +87,18 @@ static int by_file_ctrl(X509_LOOKUP *ctx, int cmd,
 int X509_load_cert_file_ex(X509_LOOKUP *ctx, const char *file, int type,
                            OSSL_LIB_CTX *libctx, const char *propq)
 {
-    int ret = 0;
     BIO *in = NULL;
-    int i, count = 0;
+    int count = 0;
     X509 *x = NULL;
 
     in = BIO_new(BIO_s_file());
 
     if ((in == NULL) || (BIO_read_filename(in, file) <= 0)) {
-        ERR_raise(ERR_LIB_X509, ERR_R_SYS_LIB);
-        goto err;
```

Comment:
There are arguments both for and against the "goto err" pattern, but nothing fundamentally wrong with the code as it was written - both `BIO_free(NULL)` and `X509_CRL_free(NULL)` are safe, even if they do nothing.

This pattern has the advantage that if anything else is added to the cleanup code then all error paths execute it.

It's common in OpenSSL, so I don't see a compelling reason to change it, unless I'm missing something

Reply:
I agree with you about `goto err`.
But IMO, at the very beginning of the function, when simple code are performed, it is better to immediately return. If add `goto err`, then have to go by code to find out what is being done under `goto err`. 
But I will fix it, because. this moment was not the target of my PR.

Reply:
Oh, if there is no cleanup to be done, a simple `return` is fine.

But most of the time there is something, even if it's just an `ERR_raise()` - TBH, I would probably have an `int err_reason` at the top of a function like this (with some useful default error value, possibly `0`), and call `ERR_raise()` in the cleanup code if `err_reason != 0`.

I'm not insisting you make this change, but seems to make things easier to read, and reduces redundancy in the `ERR_raise()` calls of repeating the `lib` value. However, it would be an increase in scope from "fix unreachable and redundant code" to "refactor" - probably better to keep things separate, and make small steps

Reply:
Your suggestion seems reasonable. But I don't remember seeing such code implementation in OpenSSL.
Do you want me to do this in the current PR?

Reply:
No, this PR is about "fix unreachable and redundant code" rather than refactoring, best to keep things small for faster review

Reply:
> But most of the time there is something, even if it's just an `ERR_raise()` - TBH, I would probably have an `int err_reason` at the top of a function like this (with some useful default error value, possibly `0`), and call `ERR_raise()` in the cleanup code if `err_reason != 0`.

This is bad pattern because it loses the line number information in error messages you get when calling ERR_raise() separately in all error cases. So no, it is better to call ERR_raise() separately.